### ISSUE TITLE
ClassCastException during web application redeploy due to the grift logging classes

### ISSUE THREAD

**Issue by meg23** [ESAPI/esapi-java-legacy#324]
_From [rkarak...@gmail.com](https://code.google.com/u/105736613281907978468/) on February 23, 2014 07:30:41_

Hi,

We are using log4j as logger implementation.
ESAPI.Logger=org.owasp.esapi.reference.Log4JLogFactory

Our log4j libraries resides in application server's lib folder. On the other hand, ESAPI jar file has to be under the web app's folder (WEB-INF/lib) since the different web app modules needs different authenticator.

When the module containing the ESAPI jar is redeployed, a ClassCastException is thrown because the log4j's LogManager was loaded by the top level class loader. Second LogManager.getLogger invocation returns the instance of org.owasp.esapi.Logger that it's class was loaded by the previous class loader.

In other words, the current ESAP architecture requires both log4j dependencies and esapi itself to be loaded by the SAME class loader.

We should move the esapi logging classes to another jar file and make esapi jar dependent on this jar.

Regards.

_Original issue: http://code.google.com/p/owasp-esapi-java/issues/detail?id=319_

**Comment by xeno6696** [comment]
This looks more like something that should be organized under the umbrella of breaking ESAPI up into several smaller libraries.  @kwwall , if there's an "epic" story/ticket to handle refactoring, this issue should be included.

**Comment by kwwall** [comment]
So for the late reply to this. I got *way* behind on me email between
Christmas and New Year's Day and this fell through the cracks.


This is really the whole point of ESAPI 3.0 (under
https://github.com/ESAPI/esapi-java). These sort of massive design changes
should, IMO, be deferred until then. My goal for ESAPI 2.x (i.e.,
https://github.com/ESAPI/esapi-java-legacy) should focus on basic bug
fixes. (There are certainly enough of them to keep us busy.)

**Comment by gnz00** [comment]
@meg23 We're having this issue as well. Did you migrate away from using ESAPI 2.x or is there a solution for hot deploys?

**Comment by xeno6696** [comment]
@gnz00 -- @meg23 is just the helper that moved the issues over from google code.  I tried [wayback](http://web.archive.org/web/20160807023033/http://code.google.com/p/owasp-esapi-java/issues/detail?id=319), but it looks like the identity of the original write will remain a mystery.

**Comment by gnz00** [comment]
@xeno6696 - Ah, thank you for clarification. ESAPI v3 appears to have stagnated in development (on Github at least), is ESAPI 2.x still the canonical OWASP Java library?

**Comment by xeno6696** [comment]
@gnz00 yeah, ESAPI 3.x was stubbed from what I recall, but in 2013 almost all ESAPI development stopped.  ESAPI 2.x is the only production-worthy offering at this time.  Project leadership will be roadmapping what the rest of the year looks like, but we should have a point-release within the month.

**Comment by kwwall** [comment]
Probably should add a "Won't fix in ESAPI 2.x" comment (like I just did :).

-kevin
